builder/simplify.cc

```diff
@@ -321,59 +327,75 @@ class simplifier : public node_mutator {
-    assert(!result_bounds.min.defined() && !result_bounds.max.defined());
+  void set_result(expr e, expr_info bounds) {
+    assert(!result_bounds.bounds.min.defined() && !result_bounds.bounds.max.defined());
     result_bounds = std::move(bounds);
     node_mutator::set_result(std::move(e));
   }
   void set_result(stmt s) {
-    assert(!result_bounds.min.defined() && !result_bounds.max.defined());
-    result_bounds = interval_expr();
+    assert(!result_bounds.bounds.min.defined() && !result_bounds.bounds.max.defined());
+    result_bounds = {interval_expr(), modulus_remainder()};
     node_mutator::set_result(std::move(s));
   }
   // Dummy for template code.
-  void set_result(stmt s, interval_expr) { set_result(std::move(s)); }
+  void set_result(stmt s, expr_info) { set_result(std::move(s)); }
 
 public:
   simplifier() {}
-  simplifier(const bounds_map& expr_bounds) : expr_bounds(expr_bounds) {}
+  simplifier(const bounds_map& bounds, const alignment_map& alignment) {
+    for (size_t ix = 0; ix < bounds.size(); ix++) {
+      var id = var(ix);
+      if (!bounds[id]) continue ;
+      expr_bounds[id] = {*bounds[id], modulus_remainder()};
+    }
+
+    for (size_t ix = 0; ix < alignment.size(); ix++) {
+      var id = var(ix);
+      if (!alignment[id]) continue;
+      if (expr_bounds[id]) {
+        expr_bounds[id]->alignment = *alignment[id];
+      } else {
+        expr_bounds[id] = {interval_expr(), *alignment[id]};
+      }
+    }
+  }
 
-  expr mutate(const expr& e, interval_expr* bounds) {
+  expr mutate(const expr& e, expr_info* bounds) {
     expr result = node_mutator::mutate(e);
     if (bounds) {
-      result_bounds = ensure_is_point(result_bounds);
+      result_bounds.bounds = ensure_is_point(result_bounds.bounds);
       if (bounds != &result_bounds) {
         *bounds = std::move(result_bounds);
       }
     } else {
-      result_bounds = {expr(), expr()};
+      result_bounds = {{expr(), expr()}, modulus_remainder()};
     }
     return result;
   }
   // Dummy for template code.
-  stmt mutate(const stmt& s, interval_expr* bounds) { return node_mutator::mutate(s); }
+  stmt mutate(const stmt& s, expr_info* bounds) { return node_mutator::mutate(s); }
   expr mutate(const expr& e) override { return mutate(e, nullptr); }
   stmt mutate(const stmt& s) override { return mutate(s, nullptr); }
 
   // When mutating a value x interpreted as boolean, we need to effectively mutate x != 0, but we can't do that directly
   // because it risks breaking the simplifiers ability to check if an expression has not changed. This helper emulates
   // this.
-  expr mutate_boolean(const expr& e, interval_expr* bounds) {
+  expr mutate_boolean(const expr& e, expr_info* bounds) {
     expr result = strip_boolean(mutate(e, bounds));
-    if (bounds) *bounds = bounds_of(static_cast<const not_equal*>(nullptr), *bounds, point(0));
+    if (bounds) bounds->bounds = bounds_of(static_cast<const not_equal*>(nullptr), bounds->bounds, point(0));
```

Comment:
I think renaming `bounds` -> `info` where appropriate would be helpful for readability here and in other places too.